File: java-demo/src/main/java/demo3/Ex1.java
package demo3;

import java.util.Arrays;
import java.util.stream.IntStream;

import akka.actor.ActorSystem;
import akka.japi.Pair;
import akka.stream.ActorMaterializer;
import akka.stream.FanInShape2;
import akka.stream.SourceShape;
import akka.stream.javadsl.FlowGraph;
import akka.stream.javadsl.Source;
import akka.stream.javadsl.Zip;
import scala.runtime.BoxedUnit;

public class Ex1 {
  public static void main(String[] args) throws Exception {
    ActorSystem system = ActorSystem.create("demo3");
    ActorMaterializer materializer = ActorMaterializer.create(system);

    Source<Integer, BoxedUnit> numbers = Source.from(() -> IntStream.range(1, 4).iterator());
    Source<Character, BoxedUnit> chars = Source.from(Arrays.asList('a', 'b', 'c'));

    Source<Pair<Integer, Character>, BoxedUnit> composite = Source.fromGraph(FlowGraph.create(b -> {
      FanInShape2<Integer, Character, Pair<Integer, Character>> zip = b.add(Zip.<Integer, Character> create());

      b.from(b.add(numbers)).toInlet(zip.in0());
      b.from(b.add(chars)).toInlet(zip.in1());
      return SourceShape.of(zip.out());
    }));

    composite.runForeach(x -> System.out.println(x), materializer);
  }
}

File: scala-demo/src/main/scala/demo3/Ex1.scala
package demo3

import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import akka.stream._
import akka.stream.scaladsl._
import scala.concurrent.duration._
import akka.stream.scaladsl.FlowGraph
import akka.stream.scaladsl.FlowGraph.Implicits._

object Ex1 extends App {
  implicit val system = ActorSystem("demo3")
  implicit val materializer = ActorMaterializer()
  import system.dispatcher

  // Create a stream composing two sources
  val numbers = Source(List(1,2,3))
  val strings = Source(List("a", "b", "c"))

  val composite = Source.fromGraph(
    FlowGraph.create() { implicit b =>
      val zip = b.add(Zip[Int,String]())
      numbers ~> zip.in0
      strings ~> zip.in1

      SourceShape(zip.out)
    }
  )

  composite.runForeach(println)
}